.github/workflows/test.yaml

```diff
@@ -35,40 +35,93 @@ jobs:
       - name: Lint
         run: pnpm lint
 
+      - name: Validate tests
+        run: pnpm test:validate
+
+  test-node:
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: CI Setup
+        uses: ./.github/actions/test-setup
```

Comment:
Calling `test-setup` here [rebuilds](https://github.com/FuelLabs/fuels-ts/actions/runs/6735982301/job/18310563161?pr=1310#step:3:129) the repo and [reruns pretest](https://github.com/FuelLabs/fuels-ts/actions/runs/6735982301/job/18310563161?pr=1310#step:4:20) even though [build](https://github.com/FuelLabs/fuels-ts/actions/runs/6735982301/job/18310373667?pr=1310#step:3:129) and [pretest](https://github.com/FuelLabs/fuels-ts/actions/runs/6735982301/job/18310373667?pr=1310#step:3:129) already ran in the `build` job. Can you share the results between jobs so that we don't repeat these steps?

Reply:
No, I believe you're mistaken. The project is not rebuilt.

The only thing happening twice seems to be the `pretest` step.

Curiously enough, it's a side-effect introduced after _your_ PR:
 - https://github.com/FuelLabs/fuels-ts/pull/1235

I didn't raise this because it is not critical, and I saw you're focused on other stuff.

Either way, this doesn't seem pertinent to this PR, which we should try to merge _ASAP_.

We can address this `pretest` cache miss in isolation later on:
 - https://github.com/FuelLabs/fuels-ts/issues/1404

Reply:
Oof, apologies. 😓 

You're right. There seems to be some duplication indeed.

Reply:
This has been introduced not due to the action in #1389, but this PR has separated build and test into 2 separate jobs. A build job precedes the test job. This has been implemented with the intention of first running a build stage, which if successful, will then trigger a test matrix stage for any number of test environments/configurations. As jobs do not share resources, this is rebuilding twice (good spot @nedsalk 🕵🏻 ). 

We can share resources between jobs by uploading the built artefact, and downloading it subsequent jobs. Which is what I will implement.

Reply:
Okay we are already using pnpm caching in CI and uploading/downloading artefacts aren't feasible here. A matrix of jobs building and testing in parallel is probably the most efficient way of doing it. However we will need a further job to amalgamate test runs and report coverage, which will require a rebuild on a fresh runner instance.

Reply:
We could introduce an internal scripts package that can then be built independently of the rest of the monorepo, for execution in CI. Which could be used to execute `nyc` in the final stage of CI.

Reply:
Oh, _now_ I see it. The rebuild occurs in _different_ flows, in parallel. In comparison, the `pretest` thing I mentioned occurs sequentially in the same flow. TBH, I don't see a problem if we build things in parallel, as it doesn't add time to the final run, but I guess it's ok to stitch things differently as long as it doesn't become overcomplex.